### PR TITLE
[WiP] Refactoring and migration to Symfony 4.1

### PR DESCRIPTION
**Pull request by gplanchat**
Refactoring various items:

* [x] #194 Refactor Images entities classes and FaceDetectProcessor
* [ ] #184 Silex end of life => migrate to Symfony 4?
* [x] Migrate the image operations to `imagine/imagine` library
* [x] Refactor the image operations to a command pattern
* [ ] Refactor the security management
* [ ] Refactor the options handling
* [x] Move the image manipulation features to `Flyimg\Image\` namespace
* [ ] Move the HTTP security and parameters handling features to `Flyimg\Http\` namespace
* [ ] Move the application-specific features to `Flyimg\App\` namespace

The document root directory has been changed to follow Symfony 4.1 recommendations, please use this config file in the Docker app:

```
# /etc/nginx/conf.d/flyimage.conf
server {
    listen  80;
    root  /var/www/html/public;

    location / {
        # try to serve file directly, fallback to index.php
        try_files $uri /index.php$is_args$args;
    }

    location ~ ^/index.php(/|$) {
        fastcgi_pass 127.0.0.1:9000;
        fastcgi_split_path_info ^(.+\.php)(/.*)$;
        include fastcgi_params;
        fastcgi_param  SCRIPT_FILENAME  $realpath_root$fastcgi_script_name;
        fastcgi_param DOCUMENT_ROOT $realpath_root;
        proxy_read_timeout 300;
    }

    error_page 504 /error_504.html;
    location = /error_504.html{
        root /var/www/html/public;
        internal;
    }
}
```

### CODE REVIEW

File: src/Image/FaceDetection/FacePositionToGeometry.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
+<?php
+
+namespace Flyimg\Image\FaceDetection;
+
+use Flyimg\Image\ShortLivedStreamedFile;
+use Imagine\Image\Box;
+use Imagine\Image\ImageInterface;
+use Imagine\Image\Point;
+use Symfony\Component\Process\Process;
+
```
Comment (baamenabar) [inline src/Image/FaceDetection/FacePositionToGeometry.php]: It will be nice to have comment explaining what is the purpose of this class

**Comment by baamenabar** [comment]
Wow! Colossal PR! 
Thanks for contributing with your work.

I´m checking superficially for now, will go deeper later today, @sadok-f  is also reviewing progress.

File: src/Image/FaceDetection/FacePositionToGeometry.php (hunk 1 above)
Comment (gplanchat) [reply src/Image/FaceDetection/FacePositionToGeometry.php]: It is the object representing a file in `imagine/imagine`, an image manipulation library abstracting imagick, gmagick and gd2 behind an unique interface.

The former `InputImage` and `OutputImage` would disappear and the logic that makes sense to keep may be moved to some other place.

**Comment by sadok-f** [comment]
Hello @gplanchat !

Thank you for this great PR and your contribution!

Here couple of points from my side:
- Can we decouple this PR to smaller once? I can see 3 major topics in the current one: Move to Symfony 4, Replace the image manipulation logic with `imagine/imagine` library and Refactoring to Command Pattern, it'll make it easier for us to review and also to make sure tests are passed and do not break the current functionalities.
- Regarding `imagine/imagine` library, I really think we should extract the logic from Flyimg and use it as a library instead, the main reason I think it has more control on the image than `imagine/imagine`, for example, the output generated mime-type/extension, it checks the browser support for `webp` and automatically generate the right image, it also generates jpg, png, gif images, also the integration with Mozjpeg (which was the main reason behind Flyimg itself)
- I really like the Command Pattern you introduced! if you can check the last issue created, we can work together to solve the issue of Input/Output classes, so the idea is to have a `BaseImage` or `StreamImage`, that we can pass it to any Processor as Input and the Output of it can be used in another Processor as Input also, using the `CommandChain` will make it easier to do that.

What do you think?

Again, thank you for your contribution and we really looking forward to get your feedback and work with you!

Sadok

**Comment by gplanchat** [comment]
Thanks for the quick review.

* I'll try to decouple the PR again, but my tries to do it took me a hard time making the current application work on Symfony 4. Maybe we could work on a `2.0.x-dev` branch and push the refactorings one by one? It may help to organize changes, eventually.
* About `imagine/imagine`, my idea was about using this library for the maximum it can provide and add what is necessary (like mozjpeg, or creating another adapter using shell commands). About the `webp` detection and other browser-related features, I did not refactor it yet in this PR and I was planning to move it to the `Flyimg\Http\` namespace, I think my description on top wasn't clear enough regarding this.
* Third point, I implemented this sort of feature at some point using file/streams/temporary resources, but removed it after a while (see https://github.com/flyimg/flyimg/pull/195/commits/4399e8c2785f72a15b0b7ac7925ac8c82122f5f3). It does not work well with `imagin/imagine`, the integration with streams wasn't straightforward and directly working with regular temporary files does create a lot of temporary files you have to clean up.

**Comment by sadok-f** [comment]
Thanks for the feedback, yes agree, let's start with `2.0.x-dev` branch.

Unfortunately, I'll be not available until next Friday, I hope by then, I'll join you working on this new branch and the new set of stuff.

**Pull request closed without merging**